packages/taiko-client/proposer/proposer.go

```diff
@@ -235,21 +236,53 @@ func (p *Proposer) fetchPoolContent(allowEmptyPoolContent bool) ([]types.Transac
 // from L2 execution engine's tx pool, splitting them by proposing constraints,
 // and then proposing them to TaikoInbox contract.
 func (p *Proposer) ProposeOp(ctx context.Context) error {
-	// Check if the preconfirmation router is set, if so, skip proposing.
-	preconfRouter, err := p.rpc.GetPreconfRouterPacaya(&bind.CallOpts{Context: ctx})
-	if err != nil {
-		return fmt.Errorf("failed to fetch preconfirmation router: %w", err)
-	}
-	if preconfRouter != rpc.ZeroAddress {
-		log.Info("Preconfirmation router is set, skip proposing", "address", preconfRouter, "time", time.Now())
-		return nil
-	}
-
 	// Wait until L2 execution engine is synced at first.
 	if err := p.rpc.WaitTillL2ExecutionEngineSynced(ctx); err != nil {
 		return fmt.Errorf("failed to wait until L2 execution engine synced: %w", err)
 	}
 
+	// Check if the preconfirmation router is set, if so, skip proposing.
+	if p.rpc.PacayaClients.PreconfRouter == nil {
+		preconfRouterAddr, err := p.rpc.GetPreconfRouterPacaya(&bind.CallOpts{Context: ctx})
+		if err != nil {
+			return fmt.Errorf("failed to fetch preconfirmation router: %w", err)
+		}
+		if preconfRouterAddr != rpc.ZeroAddress {
+			preconfRouter, err := pacayaBindings.NewPreconfRouter(preconfRouterAddr, p.rpc.L1)
+			if err != nil {
+				return fmt.Errorf("failed to create new instance of PreconfRouter: %w", err)
+			}
+			p.rpc.PacayaClients.PreconfRouter = preconfRouter
+		}
+	}
+	if p.rpc.PacayaClients.PreconfRouter != nil {
+		fallbackPreconferAddress, err := p.rpc.PacayaClients.PreconfRouter.FallbackPreconfer(&bind.CallOpts{Context: ctx})
+		if err != nil {
+			return fmt.Errorf("failed to get fallback preconfer address: %w", err)
+		}
+		if fallbackPreconferAddress != p.proposerAddress && fallbackPreconferAddress != p.ProverSetAddress {
+			log.Info("Preconfirmation is activated and proposer isn't the fallback preconfer, skip proposing",
+				"time",
+				time.Now(),
+			)
+			return nil
+		}
+		// As a fallback proposer, need to enable preconfirmation server
+		latest, err := p.rpc.L2.BlockByNumber(ctx, nil)
```

Comment:
this is gonna be a tricky one I think because in case of a reorg I think we have a race condition where when the event comes in, the chain will rewind to the first block after the last proposed block ID, and thus will be "old", and if this triggers while that happens, the fallback preconfer will propose a block, screwing everything up.

Reply:
still has same issue, i updated it.